parser.go

```diff
@@ -177,16 +178,38 @@ func (s *Session) ParseBuffer(bs []byte) (Message, error) {
 
 	sl := newSlice(bs)
 	msg.Header.unmarshal(sl)
-	msg.TemplateRecords, msg.DataRecords, err = s.readBuffer(sl)
+	readLimit := msg.Header.Length - msgHeaderLength
+	msg.TemplateRecords, msg.DataRecords, err = s.readBuffer(sl, readLimit)
 	return msg, err
 }
 
-func (s *Session) readBuffer(sl *slice) ([]TemplateRecord, []DataRecord, error) {
+// ParseBuffer extracts all message from the given buffer and returns them. Err
+// is nil if the buffer could be parsed correctly. ParseBufferAll is goroutine
+// safe.
+func (s *Session) ParseBufferAll(bs []byte) ([]Message, error) {
+	var msgs []Message
+	var err error
+
+	sl := newSlice(bs)
+
+	for sl.Len() > 0 && err == nil {
+		var msg Message
+		msg.Header.unmarshal(sl)
+		readLimit := msg.Header.Length - msgHeaderLength
+		if msg.TemplateRecords, msg.DataRecords, err = s.readBuffer(sl, readLimit); err == nil {
```

Comment:
I'd strongly prefer a more traditional `... err != nil { break }` here instead of having the positive case inside the branch.

Instead of the read limit thing I'd be happier with using `(*slice).Cut` to grab the relevant part of it into a new slice, keeping the rest in the original slice. It might be a teensy bit more code, but it keeps the (intended) property that the slice represents the full message, downstream.